plugin,etc: Rewrite to get state from Pod annotations

sharnoff opened the pull request:
> [!NOTE]
>
> Hey! This is a big PR — sorry! There's a lot of smaller commits (see below), plus one GIANT commit that does the bulk of the rewrite, building on those smaller commits.
>
> There's an overview of the changes in the commit message for the full rewrite. That might be a useful place to start to get your bearings on the changes.

---

**Commits broken down by theme:**

<details>
<summary><b>Background work: neonvm</b></summary>

1. neonvm-controller: Use non-controller owner ref for migration source (a4893d06458bf9a0f5e892885e0ca6a36dc789fe)

    > In short, there's currently no reliable way to tell whether a Pod is the source or target in a live migration without checking against the migration object itself.
    >
    > This is because the migration object initially has a unique "controller" reference on the target pod (with the source pod having a non-controlling owner reference), and once the migration is complete, the source pod is changed to have the unique controller reference setup.
    >
    > So this change is *just* to keep using a non-controller reference on the source pod after the migration completes.
    >
    > That means that we will have the following guarantees:
    >
    > * If a Pod has a "controller" owner reference for a live migration, it's  the target pod in an ongoing migration.
    > * If a Pod has a non-controller owner reference for a live migration, it's the source pod in a maybe-ongoing migration.
    > * If the Pod has a "controller" owner reference for a virtual machine, it's the current source of the VM. So if a migration source pod *doesn't* have an owner reference for the VM, then the migration has completed and there is no longer a use for the old source pod.

2. neonvm: Add helpers to get pod ownership (3e674b58c79f85185dcba791996cf84346d10350)

    > We had some existing, similar helpers in `pkg/util`, but I figured it makes more sense for those to be defined in `neonvm/apis/...`.
    >
    > This change originated while working on the stateless scheduler prototype, where I found myself wanting a reliable way to determine what role a pod has in a live migration from only the pod's metadata.

3. neonvm-controller: Update runner pod metadata while pending (3086c594df00502d40db06d07375bee09ccda1e3)

    > We weren't doing this previously, and it means that propagation of labels/annotations can be arbitrarily delayed while the VM starts.

</details>

<details>
<summary> <b>Background work: <code>util/watch</code></b> </summary>

4. util/watch: Store `HandlerFuncs[*T]` in `Store[T]` (8a9cfc70e3b367bb03bd9e7f846f017a4bdd8179)

    > Also, change `handleEvent[T any, P ~*T]()` from a function to a method on `*Store[T]`, now that we have the handlers translated from `P` → `*T`.
    >
    > This opens up a lot of ways to make the code cleaner, and the handlers part in particular is required to implement `(*Store[T]).NopUpdate()` in a later commit.

5. util/watch: Add `(*Store[T]).NopUpdate()` method (a757fcdc0db6796d874ceb85591991ee0957d837)

    > There's some info about this in the added comment.
    >
    > tl;dr: We need it for the stateless scheduler work to be able to re-inject items into the reconcile queue while maintaining the watch as the source of truth for what each object is.

6. util/watch: Override GVK on incoming objects (426070a351ec12ce2bb6df2b4a66b4dc34965b18)

    > The K8s API server and client-go together have the behavior that objects returned from `List()` calls do not have `TypeMeta` set.
    >
    > For one-off `List()` requests this is fine becuase you already know the type! But this interacts poorly with the generated implementations of objects' `.GetObjectKind().GroupVersionKind()`, as those just directly read from the `TypeMeta` fields (which again: are not set).
    >
    > So this commit works around this behavior by getting the GVK at the start of the `Watch()` call and explicitly setting it on all incoming objects.

7. util/watch: Add `(*Store[T]).Listen()` method (9e9f4406f8d0867892f57439b435afd2d282bb3f)

    > The Listen() method returns a `util.BroadcastReceiver` that will be updated whenever the object is modified or deleted.
    >
    > This is required *for now* for the stateless scheduler work, so that we can be separately notified when there's changes to an object without hooking deeper into the internal state.
    >
    > We can probably remove this once the scheduler plugin's agent request handler server is removed, at the end of the stateless scheduler work.

</details>

<details>
<summary><b>The one big commit doing the rewrite</b></summary>

8. plugin: Rewrite to get state from Pod annotations (cf468d8f4a8e94694aecfc174f33bf536d76bb1b)

    > a.k.a. "Stateless Scheduler".
    >
    > This is effectively a full rewrite of the scheduler plugin. At a high level, the existing external interfaces are preserved:
    >
    > - The scheduler plugin still exposes an HTTP server for the autoscaler-agent (for now); and
    > - The scheduler plugin is still a plugin.
    >
    > However, instead of storing the state for approved resources in-memory, in the scheduler plugin, we now treat annotations on the Pod as the source of truth for requested/approved resources.
    >
    > A brief overview of the internal changes to make this happen:
    >
    > 1. The state of resource reservations can be constructed entirely from Node and Pod objects. We *do* store that, and update as objects change, but it's only for convenience and not a strict requirement.
    >    
    >    One tricky piece is with scheduling. For that, we store a set of pods that have gone through the plugin methods but haven't actually had the `spec.nodeName` field set.
    >    
    >    For more info, the `pkg/plugin/state` package contains all the pure logic for manipulating resources.
    >
    > 2. Each watch event on Node/Pod objects is now placed into a "reconcile" queue similar to the controller framework. Reconcile operations are a tuple of (object, event type, desired reconcile time) and are retried with backoff on error/panic.
    >    
    >    For a detailed look, the 'pkg/plugin/reconcile' package defines the reconcile queue and all its related machinery.
    >
    > 3. The handler for autoscaler-agent requests no longer accesses the internal state and instead directly patches the `VirtualMachine` object to set the annotation for requested resources, and then waits for that object to be updated.
    >    
    >    Once the autoscaler-agent is converted to read and write those annotations directly, we will remove the HTTP server.
    >
    > 4. `pkg/util/watch` was changed to allow asking to be notified when
   there's changes to an object, via the new `(*Store[T]).Listen()` API.
    >    
    >    This was required to implement (3), and can be removed once (3) is no longer needed, if it doesn't become used in the autoscaler-agent.
    >
    > 5. `pkg/util/watch` was changed to allow triggering no-op update events, which - for our usage - will trigger requeuing the object. This solves two problems:
    >    
    >    1. During initial startup, we need to defer resource approval until all Pods on the Node have been processed -- otherwise, we may end up unintentionally overcommitting resources based on partial information.
    >      
    >       So during startup, we track the set of Pods with deferred approvals, and then requeue them all once startup is over by triggering no-op update events in the watch store.
    >    
    >    2. Whenever we handle changes for some Pod, it's worthwhile to handle certain operations on the Node -- e.g., triggering live migration if the reserved resources are too high.
    >    
    >       While we *could* do this as part of the Pod reconciling, we get more fair behavior (and, better balancing under load) by instead triggering re-reconciling the Pod's Node.
    >    
    >    Why can't this be done elsewhere? In short, consistency. Fundamentally we need to use a consistent view of the object that we're reconciling (else, it might not be no-op), and the source of truth for the current value of an object *within the scheduler plugin* is the watch store.

</details>

---

**Remaining TODOs:**

- [ ] Flesh out the reconcile metrics
- [ ] Add metrics for k8s CRUD operations (mainly: rate of VirtualMachine update operations)
- [ ] Test node metrics
- [ ] Test that deferred reconciles on startup work correctly
- [ ] Test that setting VM CPU/memory use fields and enabling autoscaling at the same time works correctly.

**Open questions:**

- Which commits should be squashed, or extracted into separate PRs?

sharnoff commented:
* Squashed 06240ddcbf9e2ef8ea46f391a5ee0589734a8dca..0b64f0832ac6112cca1313bd1a7d52b27b71b97b ([diff](https://github.com/neondatabase/autoscaling/compare/06240ddcbf9e2ef8ea46f391a5ee0589734a8dca..0b64f0832ac6112cca1313bd1a7d52b27b71b97b))
* Rebased onto main: 0b64f0832ac6112cca1313bd1a7d52b27b71b97b..6c76e9d60fd5dfca5ed503b322b11a01d5e5190a [(diff)](https://github.com/neondatabase/autoscaling/compare/0b64f0832ac6112cca1313bd1a7d52b27b71b97b..6c76e9d60fd5dfca5ed503b322b11a01d5e5190a)

Omrigan reviewed (commented):
Partial review, got to the middle of handle_pod.

petuhovskiy commented:
> Which commits should be squashed, or extracted into separate PRs?

The first 7 commits look quite isolated from the big "stateless scheduler" commit. They also LGTM, so you can create a PR out of these commits and merge them to main, if you want.

petuhovskiy commented on pkg/plugin/handle_node.go:
```diff
@@ -0,0 +1,169 @@
+// This is typically used to force nodes to stay up-to-date after we update a pod on the node, while
+// helping with fairness between time spent reconciling the pod vs the node.
+func (s *PluginState) updateNodeMetricsAndRequeue(logger *zap.Logger, ns *nodeState) {
+	if err := s.requeueNode(ns.node.Name); err != nil {
```
question: Can we have a deadlock here? No concrete example from me, just noticed that this call requires some mutexes to be held:
- caller takes `PluginState.mu`
- we're still holding it inside `updateNodeMetricsAndRequeue()`
- `requeueNode` uses `indexedNodeStore.Store.mutex` during lookup
- `nodeWatchStore.NopUpdate` takes `nodeWatchStore.mutex`
- and calls `queue.Enqueue`

And trying to answer my question here:
- I guess `queue` is fully async and doesn't call any handlers, it just sends notifications to workers. So it shouldn't be possible to deadlock a queue.

sharnoff replied:
> `queue` is fully async and doesn't call any handlers, it just sends notifications to workers. So it shouldn't be possible to deadlock a queue.

Exactly that, yeah. Although this locking is quite sketchy... Any ideas for improvement?

petuhovskiy replied:
Was just thinking out loud, no specific ideas for improvement. I'm afraid we can have some locking issue (deadlock, etc) in theory, but don't think we should do anything ahead of time.

petuhovskiy reviewed (commented):
Finished my first reading, overall the code looks very good. Will come back for a second iteration next year :) 

Made some small refactorings along the way, mostly it is about moving metrics-related code to a new package: https://github.com/neondatabase/autoscaling/pull/1182/commits

The changes are described in commit messages, feel free to accept/reject.

sharnoff commented on pkg/plugin/state/node.go:
```diff
@@ -0,0 +1,424 @@
+func (r *NodeResources[T]) add(p *PodResources[T], migrating bool) {
+	r.Reserved += p.Reserved
+	if migrating {
+		r.Reserved += p.Reserved
```
Oh, great catch. That should be adding to `Migrating`, not `Reserved` 🤦

sharnoff replied:
Fixed in 37482d4921a8386f00eb107c9d012245597b1b74

sharnoff commented on pkg/plugin/state/xactmap.go:
```diff
@@ -0,0 +1,137 @@
+// If forEach returns (*LoopControl).Break(), iteration will end early.
+//
+// Deleting elements from the map during iteration is always sound. They will not be visited later.
+func (m *XactMap[K, V]) Iter(forEach func(LoopControl, K, V) LoopControlFlow) {
```
Ah, nice - TIL! Implemented in 3e24629ddf372fba057b6a3404034d02cc925607

sharnoff commented on pkg/plugin/handle_pod.go:
```diff
@@ -0,0 +1,548 @@
+	}
+}
+
+func (s *PluginState) updatePod(
```
Tackled that in 71fc62f596e55bb23f4d9e1ec2f6326dcc3354e7 — there's actually two different actions we might take after unlock, so I figured returning a function to execute is simpler.

sharnoff commented on neonvm/apis/neonvm/v1/pod_helpers.go:
```diff
@@ -0,0 +1,94 @@
+	return extractFromAnnotation[VirtualMachineResources](pod, VirtualMachineResourcesAnnotation)
+}
+
+func extractFromAnnotation[T any](pod *corev1.Pod, annotation string) (*T, error) {
```
Hm, if we put it in `pkg/util`, we get an import cycle, but I don't think it belongs in either `pkg/api` or `neonvm/apis/neonvm/v1`...

I'm inclined to leave it as-is for now, but LMK what you think

sharnoff replied:
Ack; leaving as-is.

sharnoff commented on pkg/plugin/globalstate.go:
```diff
@@ -0,0 +1,193 @@
+	}
+}
+
+func recordK8sOp(metrics pluginMetrics, opKind string, objKind string, objName string, err error) {
```
This was implemented in b5f3307e44b8455c980e24c0a3771fa91e2cd836

sharnoff commented on pkg/plugin/reconcile/middleware.go:
```diff
@@ -0,0 +1,326 @@
+
+// Middleware wraps a reconcile operation to insert its own logic
+type Middleware interface {
+	Call(*zap.Logger, MiddlewareParams, MiddlewareHandlerFunc) (Result, error)
```
The reason I didn't go with that approach is because storing the handler in the middleware requires either:

1. Putting together all the middleware requires explicit ordering of how we construct them, which risks making the code harder to refactor (because you can't "just" move things around, due to the web of dependencies); OR
2. Collecting the set of middleware as a set of `func(handler) Middleware` to be assembled at the end

I figured that providing the handler func in the call to middleware allows it to be more isolated, which -- for me, at least -- reduces the number of things I need to hold in my head when thinking about how it works.

Does that make sense? If so, WDYT?

sharnoff replied:
Also worth noting that the approach you mentioned requires a single handler, rather than the distinct handler for each type that the queue currently receives as input -- discussed that here: https://github.com/neondatabase/autoscaling/pull/1163#discussion_r1901208166

sharnoff commented on pkg/plugin/reconcile/options.go:
```diff
@@ -0,0 +1,106 @@
+	apply func(*queueSettings)
+}
+
+// queueSettings is the internal, temporary structure that we use to hold the results of applying
```
I considered doing the construction of the default middleware outside `reconcile.NewQueue`, but ultimately decided not to, because:

1. These particular pieces of middleware (logging, error backoff, panic handling) should always be included — it would be an error to not include them, because the behavior of the queue would be open to easily preventable risks (busy-looping on an error, panicking taking down the process, etc.). So, it's _probably_ better from an API perspective to make sure that they must always be included!
2. The functions that they can be configured with _are_ optional ­— e.g., they may only be used for incrementing metrics, so IMO it's more natural to have the API express that these are optional, even if in practice we always use them

Does that make sense? WDYT?

sharnoff commented on pkg/plugin/reconcile/queue.go:
```diff
@@ -0,0 +1,449 @@
+
+		// Check that this isn't a duplicate
+		if _, ok := handlersByType[gvk]; ok {
+			return nil, fmt.Errorf("Duplicate handler for object type %T with GVK %q", obj, fmtGVK(gvk))
```
Presumably staticcheck style lints aren't enabled? The one you're referring to seems to be [ST1005](https://staticcheck.dev/docs/checks#ST1005).

Across the rest of the codebase, of the calls to `fmt.Errorf` that start with a letter, about 1/3 are capitalized and 2/3 are lowercase. I'm happy to make that change separately, but it's broader than just this PR.

sharnoff replied:
Switched to lowercase errors in f1f164df763dcaa02a165737874c00e87ac2a093

sharnoff commented on pkg/plugin/reconcile/middleware.go:
```diff
@@ -0,0 +1,326 @@
+	Call(*zap.Logger, MiddlewareParams, MiddlewareHandlerFunc) (Result, error)
+}
+
+// MiddlewareHandlerFunc is an enriched version of HandlerFunc that accepts more parameters, so that
```
Mostly, I figured it was cleaner to have a simpler `HandlerFunc` — `MiddlewareParams` exists because it's easier for all the middleware to have access to common object attributes (e.g., GVK) than trying fallible extraction in each one, whereas the end handler probably doesn't need those.

sharnoff replied:
Fair enough, renamed to `ObjectParams` in 987d73bf216e7eb05a0559fede928a63228c0112

sharnoff commented on pkg/plugin/handle_pod.go:
```diff
@@ -0,0 +1,548 @@
+	var ok bool
+	ns, ok = s.nodes[nodeName]
+	if !ok {
+		return fmt.Errorf("pod's node %q is not present in local state", nodeName)
```
TBH that seems like it'd add a lot of complexity that I'm not sure is worth it — adding a pod potentially has to add a placeholder node but do nothing else, and deleting a pod potentially has to also delete the node it's on, if the node was already deleted and the pod is the last one.

I'm OK with complexity encapsulated into the `state` package because that's easy to unit-test, but I don't currently see a clean interface boundary we can draw here without unit-testing the entire global state (which is _maybe_ fine, but a lot of effort! that's what the `state` package is meant to help with)

Maybe I'm missing something? WDYT?

sharnoff commented on pkg/plugin/globalstate.go:
```diff
@@ -0,0 +1,193 @@
+
+	metrics pluginMetrics
+
+	requeuePod      func(uid types.UID) error
+	requeueNode     func(nodeName string) error
```
To answer what these IDs are, [in short](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/):

> Each [object](https://kubernetes.io/docs/concepts/overview/working-with-objects/#kubernetes-objects) in your cluster has a [Name](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names) that is unique for that type of resource. Every Kubernetes object also has a [UID](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids) that is unique across your whole cluster.

As for why these two `requeueFoo` functions take two different IDs:

- Pods can be trivially recreated with the same name. So in this PR, we always include the UID whenever dealing with a Pod object, and only ever store maps of them keyed by UID.
- while Nodes _can_ have duplicate names at different times, we also have to deal with the fact that Pods _only_ refer to nodes by name, not UID — see the `.spec.nodeName` field. We already need to have some way of doing node name → node state lookups to handle pod events, so for consistency this PR only ever refers to nodes by _name_ rather than by UID.

sharnoff commented:
- Squashed 71fc62f596e55bb23f4d9e1ec2f6326dcc3354e7..11b50acaa35519ac5b598fe3e7ef75cdd25b6986 ([diff](https://github.com/neondatabase/autoscaling/compare/71fc62f596e55bb23f4d9e1ec2f6326dcc3354e7..11b50acaa35519ac5b598fe3e7ef75cdd25b6986))
- Rebased 11b50acaa35519ac5b598fe3e7ef75cdd25b6986..d04ce1b36113d72ad9e6bf459f502f4f9602f895 ([diff](https://github.com/neondatabase/autoscaling/compare/11b50acaa35519ac5b598fe3e7ef75cdd25b6986..d04ce1b36113d72ad9e6bf459f502f4f9602f895))

sharnoff commented on pkg/plugin/reconcile/event.go:
```diff
@@ -0,0 +1,51 @@
+		return k
+	}
+
+	if k == EventKindAdded && other == EventKindDeleted || k == EventKindDeleted && other == EventKindAdded {
```
> Does it mean that events were reordered (because delete cannot happen before add)?

Exactly, yeah

> Also, what happens if object is deleted and then instantly recreated? Will it always have a different UID?

Yep, k8s guarantees that UIDs are unique across the lifetime of a cluster — more on that here: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids

sharnoff commented on pkg/neonvm/controllers/vmmigration_controller.go:
```diff
@@ -451,7 +451,12 @@ func (r *VirtualMachineMigrationReconciler) Reconcile(ctx context.Context, req c
 			}
 
-			// Redefine ownerRef for the target Pod
+			// Redefine owner references for the source and target pod:
```
Reworded that commit as suggested, and opened #1194 as a separate PR.

petuhovskiy commented on pkg/plugin/framework_methods.go:
```diff
@@ -0,0 +1,610 @@
+	if !ok {
+		msg := "Node not found in local state"
+		logger.Error(msg)
+		return framework.NewStatus(framework.Error, fmt.Sprintf("%s: %s", msg, err.Error()))
```
issue (blocking): `err` is incorrectly used and is always `nil` here

suggestion:
```suggestion
		return framework.NewStatus(framework.Error, msg)
```

sharnoff replied:
Good catch, thanks! Fixed in 3a9f02cb6ce1aecb1193f9ba31cb00ad8aa085af

petuhovskiy commented on pkg/plugin/framework_methods.go:
```diff
@@ -0,0 +1,610 @@
+) (_ int64, status *framework.Status) {
+	ignored := e.state.config.ignoredNamespace(pod.Namespace)
+
+	e.metrics.IncMethodCall("NormalizeScore", pod, ignored)
```
issue (blocking): `NormalizeScore` here looks like a copy-paste typo, I guess it should be `Score`?

suggestion: replace `NormalizeScore` with `Score` on lines 287, 289, 293

sharnoff replied:
Good catch, thanks! b47ba954b898ee519e7cc82b8d5f301b101b2ebe

petuhovskiy commented on pkg/plugin/framework_methods.go:
```diff
@@ -0,0 +1,610 @@
+	if !ok {
+		msg := "Node not found in local state"
+		logger.Error(msg)
+		return framework.NewStatus(framework.Error, fmt.Sprintf("%s: %s", msg, err.Error()))
```
issue (blocking): `err` is incorrectly used and is always `nil` here

suggestion:
```suggestion
		return framework.NewStatus(framework.Error, msg)
```

thought: This is a second case of this issue in the file, I think we should enable linter to catch similar issues. ChatGPT said that `staticcheck` should catch issues like this, not sure why it didn't see any issue here (we already have it enabled).

sharnoff replied:
Yeah, +1 that this should be caught by a linter. I'm not sure why staticcheck isn't already catching it...

(fixed in 3a9f02cb6ce1aecb1193f9ba31cb00ad8aa085af)

petuhovskiy commented on pkg/plugin/framework_methods.go:
```diff
@@ -0,0 +1,610 @@
+	if !ok {
+		msg := "Node not found in local state"
+		logger.Error(msg)
+		status := framework.NewStatus(framework.Error, fmt.Sprintf("%s: %s", msg, err.Error()))
```
issue (blocking): `err` is incorrectly used and is always `nil` here

suggestion:
```suggestion
		status := framework.NewStatus(framework.Error, msg)
```

sharnoff replied:
Fixed in 3a9f02cb6ce1aecb1193f9ba31cb00ad8aa085af

petuhovskiy reviewed (approved):
Found a couple of small issues. I can approve everything except `framework_methods.go`, will take a look at it again tomorrow.

sharnoff commented on autoscale-scheduler/cmd/main.go:
```diff
@@ -25,6 +27,7 @@ import (
 func main() {
 	logConfig := zap.NewProductionConfig()
 	logConfig.Sampling = nil // Disable sampling, which the production config enables by default.
+	logConfig.DisableStacktrace = true
```
I think it was because the error logs from failed reconciles were spamming the logs with repeats of the same stack trace (because those errors are generated by the logging middleware, which is always called the same way).

Added a comment to clarify this in 895cb6c6cd51b92e25a4575755076e6ec79a641a

sharnoff commented on pkg/plugin/setup.go:
```diff
@@ -0,0 +1,70 @@
+package plugin
+
```
Makes sense, yeah -- moved in 4f0f84c869c65415c456ace106c1d57af40d69fb

sharnoff commented on pkg/plugin/state/node.go:
```diff
@@ -0,0 +1,420 @@
+		newState.CPU.Watermark != n.CPU.Watermark || newState.Mem.Watermark != n.Mem.Watermark
+
+	// Propagate changes to labels:
+	for label, value := range n.Labels.Entries() {
```
Oof, good catch, thanks. Fixed and added tests in 2c5a1f4c1549abd622fd6b4c83cc077fce18d3e5.

sharnoff commented on pkg/plugin/framework_methods.go:
```diff
@@ -0,0 +1,610 @@
+
+	ns, ok := e.state.nodes[nodeName]
+	if !ok {
+		logger.Error("Node not found in local state", logFieldForNodeName(nodeName))
```
Good catch, thanks! Fixed in 2159c5a71e6297c8ac30897505f5b18c124eca5a

sharnoff commented on pkg/plugin/reconcile/middleware.go:
```diff
@@ -0,0 +1,326 @@
+	maxErrorWait     = time.Minute
+)
+
+func NewErrorBackoffMiddleware(typs []schema.GroupVersionKind, callback ErrorStatsCallback) *ErrorBackoffMiddleware {
```
Good point -- added some comments in 764f95161adf8bef9b34babff12a6cba38ea1bce

Omrigan reviewed (approved):
Almost done!

sharnoff commented on pkg/plugin/reconcile/event.go:
```diff
@@ -0,0 +1,51 @@
+	EventKindModified EventKind = "Modified"
+	EventKindDeleted  EventKind = "Deleted"
+
+	// EventKindEphemeral represents the combination when an addition was not handled before the
+	// object was deleted, yet we still may want to process either of these.
```
Ack, fair enough

sharnoff commented:
Updates:

* Small changes from review: 86376ea32391d60ea33f404844344c37b11044f4..987d73bf216e7eb05a0559fede928a63228c0112 ([diff](https://github.com/neondatabase/autoscaling/compare/86376ea32391d60ea33f404844344c37b11044f4..987d73bf216e7eb05a0559fede928a63228c0112))
* Squashed 987d73bf216e7eb05a0559fede928a63228c0112..289735a08a84cd24ca2302296e263daa56c546ca ([diff](https://github.com/neondatabase/autoscaling/compare/987d73bf216e7eb05a0559fede928a63228c0112..289735a08a84cd24ca2302296e263daa56c546ca))
* Rebased onto main 289735a08a84cd24ca2302296e263daa56c546ca..dcfb84277864f378b62a7f3b6f19fe9a976d5c18 ([diff](https://github.com/neondatabase/autoscaling/compare/289735a08a84cd24ca2302296e263daa56c546ca..dcfb84277864f378b62a7f3b6f19fe9a976d5c18))

sharnoff commented:
Updates:

- Rebased onto main dcfb84277864f378b62a7f3b6f19fe9a976d5c18..24b5ff5a583fce6131bf1538a6142832d857e5e6 ([diff](https://github.com/neondatabase/autoscaling/compare/dcfb84277864f378b62a7f3b6f19fe9a976d5c18..24b5ff5a583fce6131bf1538a6142832d857e5e6))

sharnoff commented:
Updates:

- Small review fix 24b5ff5a583fce6131bf1538a6142832d857e5e6..f1f164df763dcaa02a165737874c00e87ac2a093 ([diff](https://github.com/neondatabase/autoscaling/compare/24b5ff5a583fce6131bf1538a6142832d857e5e6..f1f164df763dcaa02a165737874c00e87ac2a093))
- Squash f1f164df763dcaa02a165737874c00e87ac2a093..d1a208bc6e32d0a670eb6fa8ff78af8bb9d4fdeb ([diff](https://github.com/neondatabase/autoscaling/compare/f1f164df763dcaa02a165737874c00e87ac2a093..d1a208bc6e32d0a670eb6fa8ff78af8bb9d4fdeb))
- Rebase onto main d1a208bc6e32d0a670eb6fa8ff78af8bb9d4fdeb..76e0e9229d82e71fa9f2a29ae367158968b16a28 ([diff](https://github.com/neondatabase/autoscaling/compare/d1a208bc6e32d0a670eb6fa8ff78af8bb9d4fdeb..76e0e9229d82e71fa9f2a29ae367158968b16a28))
